### DISCUSSION TITLE
[observation-beta] Expected to receive binding but it didn't

### DISCUSSION THREAD

**Discussion by x-0o0** [pointfreeco/swift-composable-architecture#2686] [Beta]
First of all, here is my very simple swift package that includes 2 swift files: one for defining the reducer and the other one is for testing the reducer

[Experiment.zip](https://github.com/pointfreeco/swift-composable-architecture/files/13783316/Experiment.zip)

Here's the simplified reducer code:
This reducer describes that the user can select the items to delete after tapped edit button and then delete items by tapping delete button. Also, if there's no selected item, delete button should be disabled.

```swift
  @ObservableState
  struct State: Equatable {
    /// current items
    var items: IdentifiedArray<Item.ID, Item> = []
    /// The IDs for deletion
    var selectedItemIDs: Set<Item.ID> = []
    var editMode: EditMode = .none
    
    enum EditMode: Equatable {
      case `none`
      case editing(deletable: Bool)
    }
    
    init(items: [Item]) {
      self.items = IdentifiedArray(uniqueElements: items)
    }
  }

  var body: some ReducerOf<Self> {
    BindingReducer()
    
    Reduce { state, action in
      switch action {
      case .binding(\.selectedItemIDs):
        /// with `selectedItemIDs`, update the status whether it's deletable or not
        switch state.editMode {
        case .editing:
          state.editMode = .editing(deletable: !state.selectedItemIDs.isEmpty)
          return .none
        default:
          return .none
        }
        
      case .editButtonTapped:
        /// enable edit mode
        state.editMode = .editing(deletable: false)
        return .none
        
      case .deleteButtonTapped:
        /// remove selected items
        state.selectedItemIDs.forEach {
          state.items.remove(id: $0)
        }
        /// reset `selectedItemIDs` 
        state.selectedItemIDs.removeAll()
        return .none
        
      case .binding:
        return .none
      }
    }
  }
```

From my test code, I expected `binding(\.selectedItemIDs)` action to be called  when I send `.deleteButtonTapped` via store, because it resets the `selectedItemIDs` when receives the `.deleteButtonTapped` action. However it doesn't.
Why it doesn't invoke the binding action?

```swift
    await store.send(
      .binding(.set(\.selectedItemIDs, [items[0].id]))
    ) {
      /// updated by binding action
      $0.editMode = .editing(deletable: true)
    }
    
    /// tapped delete button
    await store.send(.deleteButtonTapped) {
      $0.selectedItemIDs = [] // Expected to receive binding action with the empty `selectedItemIDs`
      $0.editMode = .editing(deletable: false) // 🛑 Actual value is `.editing(deletable: true)`
    }
```

**Comment by mbrandonw** [comment]
Hi @x-0o0, the test failure is correct, and I think you may have some confusion about how binding actions work.

When you send the `deleteButtonTapped` action you empty out the `selectedItemIDs` state, but that doesn't cause a `.binding(\.selectedItemIDs)` action to be sent. Changes to state inside the reducer never "magically" send binding actions.

If you want to execute the logic inside `.binding(\.selectedItemIDs)` in multiple places of the reducer you can extract it out to a helper method, and then invoke it anywhere you want.

**Reply by x-0o0** [reply]
@mbrandonw Really Thanks! I realized how binding actions work.
Happy new year!
